tests/test_suggestion.c

```diff
@@ -0,0 +1,401 @@
+    match_t **set = calloc(1, sizeof(match_t*));
+
+    trie_insert_string(t, "ap");
+
+    int rc = suggestions(set, t, "p", "a", 1, 1);
+
+    cr_assert_eq(0, rc, "suggestions() failed");
+    cr_assert_eq(strcmp(set[0]->str, "ap"), 0, "try_swap string failed");
+    cr_assert_eq(set[0]->edits_left, 0, "try_swap value failed");
+}
+
+// Test a double swap 
+Test(suggestion, swap_s1) {
+    trie_t *t = trie_new('\0');
+    match_t **set = calloc(5, sizeof(match_t*));
+
+    trie_insert_string(t, "4uiwifnw");
+
+    int rc = suggestions(set, t, "", "4uiwnfiw", 2, 5);
+
+    cr_assert_eq(0, rc, "suggestions() failed");
+    cr_assert_eq(strcmp(set[0]->str, "4uiwifnw"), 0, "try_swap string failed");
+    cr_assert_eq(set[0]->edits_left, 0, "try_swap value failed");
+}
+
+// Test a simple insert
+Test(suggestion, insert_s0) {
+    trie_t *t = trie_new('\0');
+    match_t **set = calloc(1, sizeof(match_t*));
+
+    trie_insert_string(t, "+");
+
+    int rc = suggestions(set, t, "", "", 1, 1);
+
+    cr_assert_eq(0, rc, "suggestions() failed");
+    cr_assert_eq(strcmp(set[0]->str, "+"), 0, "try_swap string failed");
+    cr_assert_eq(set[0]->edits_left, 0, "try_swap value failed");
+}
+
+// Test a double insert
+Test(suggestion, insert_s1) {
+    trie_t *t = trie_new('\0');
+    match_t **set = calloc(5, sizeof(match_t*));
+
+    trie_insert_string(t, "fnjk43?>.f@");
+
+    int rc = suggestions(set, t, "", "fnk43?>.f", 2, 5);
+
+    cr_assert_eq(0, rc, "suggestions() failed");
+    cr_assert_eq(strcmp(set[0]->str, "fnjk43?>.f@"), 0, "try_swap string failed");
+    cr_assert_eq(set[0]->edits_left, 0, "try_swap value failed");
+}
+
+/*************   These functions test suggestions() as a whole   *************/
+
+// Test regular suggestion generation 1
+Test(suggestion, suggestions_s0) {
+    trie_t *t = trie_new('\0');
+    match_t **set = calloc(4, sizeof(match_t*));
+
+    trie_insert_string(t, "cops");
+    trie_insert_string(t, "8op");
+    trie_insert_string(t, "op");
+    trie_insert_string(t, "ocp");
+
+    int rc = suggestions(set, t, "", "cop", 1, 4);
+    cr_assert_eq(0, rc, "suggestions() failed");
+
+    // Get the items in order since suggestions() doesn't sort
+    qsort(set, 4, sizeof(match_t*), cmp_match);
+
+    cr_assert_eq(0, strncmp(set[0]->str, "8op", MAXLEN), 
+                "suggestions() first result incorrect");
+    cr_assert_eq(0, strncmp(set[1]->str, "cops", MAXLEN), 
+                "suggestions() second result incorrect");
+    cr_assert_eq(0, strncmp(set[2]->str, "ocp", MAXLEN), 
+                "suggestions() third result incorrect");
+    cr_assert_eq(0, strncmp(set[3]->str, "op", MAXLEN), 
+                "suggestions() fourth result incorrect");
+}
+
+// Test regular suggestion generation 1
+Test(suggestion, suggestions_s1) {
+    trie_t *t = trie_new('\0');
+    match_t **set = calloc(5, sizeof(match_t*));
+
+    trie_insert_string(t, "stall");
+    trie_insert_string(t, "satll"); // success but not in list
```

Comment:
What do you mean by "success but not in list"?

Reply:
@dwahme

Reply:
Comment updated to:

```
    trie_insert_string(t, "satll"); 
    // Although the above item is within the correct matches, there isn't enough space for it
    // since there are too many other matches and it's after the matches alphabetically
```

I think that should suffice, but please let me know if you need any more clarification on this.